src/Build/Instance/ProjectInstance.cs

```diff
@@ -561,9 +561,13 @@ internal ProjectInstance(Evaluation.Project.Data data, string directory, string
             this.CreateTargetsSnapshot(data.Targets, data.DefaultTargets, data.InitialTargets, data.BeforeTargets, data.AfterTargets);
             this.CreateImportsSnapshot(data.ImportClosure, data.ImportClosureWithDuplicates);
 
-            this.Toolset = data.Toolset; // UNDONE: This isn't immutable, should be cloned or made immutable; it currently has a pointer to project collection
             this.SubToolsetVersion = data.SubToolsetVersion;
-            this.TaskRegistry = data.TaskRegistry;
+            this.TaskRegistry = data.TaskRegistry?.DeepClone();
+
+            // If the task registry uses toolset identical to the one in data instance - deep clone it just once.
```

Comment:
Please explain why - in comment.  The fact that it is about reusing already cloned instance by line above might  not be obvious for reader.